### DISCUSSION TITLE
Writing symbols via berry scripting console

### DISCUSSION THREAD

**Discussion by waltkind** [arendst/Tasmota#20277] [Q&A]
from the sample file I use the bottom buttons.
`{"id":101,"obj":"btn","x":0,"y":295,"w":80,"h":25,"action":"prev","bg_color":"#1fa3ec","radius":10,"border_side":1,"text":"\uF053","text_font":"montserrat-20"}`.
When I enter `p0b101.text='15'` in the console, the text changes.
W.hen I enter `p0b101.text='\uF015'` the I get `BRY: Exception> 'syntax_error' - input:1: invalid octal number`

What is wrong with this?  And is there a list of valid symbol codes for HASPmota?

**Comment by sfromis** [comment]
The error message you get is not related to HASPmota, as you get the same error from `\uF015` when trying to do it as a simple assignment of a plain variable. Your expectation of Unicode-like hex escapes with `\u` is not how it works, as using `\` is supposed to be for one of the valid escape sequences:
https://github.com/berry-lang/berry/wiki/Chapter-1#string-literal

That section also describes how valid escape sequences work for "any character", using `\xF0` or `\360` for the code point you want; the latter syntax with octal numbers being the default for escapes (unless having a valid letter for something built-in) is also the reason for the unintuitive message you get when not using a valid letter.

**Reply by waltkind** [reply]
Why does it work in pages.jsonl when it is not the right code?

**Reply by sfromis** [reply]
The file `pages.jsonl` has nothing to do with how strings are written in the Berry language; it is instead JSON, and thus has to match the escaping rules of the JSON specification.
https://www.json.org/json-en.html

This includes the syntax where a string can contain escape codes with `\` and the next character being a letter, where some of these for obvious reasons are the same abbreviation as used in Berry, but not including `\u` with 4 hex digits, where Berry instead has `\x` with two hex digits, or no letter for octal.

**Reply by waltkind** [reply]
Thank you very much for explaining this. I was not aware of that difference.
But is there a list of valid codes for HASPmota to use in pages.jsonl?

**Reply by sfromis** [reply]
AFAICT, what is valid in HASPmota is simply what is valid in LVGL, and what is valid in LVGL is simply what is valid in the universal display driver, and what is valid there is simply what's valid in the font you use.

Thus, know your font :grin:
